### PR TITLE
Added dashboard with home page, station performance dashboard, and mock subscription page

### LINKED ISSUE #55: [Create] Dashboard prototype
Use streamlit to create a prototype for the dashboard and mock up some data to experiment with how we could present it.

User Story:
- As a developer, a prototype will let me experiment with what is feasible and get a solid picture of what my final product could look like.

### PR DESCRIPTION
**Pull request by adamski201**
This PR adds a first draft of the dashboard.

Home.py contains the formatting and text content for the home page.

1_Station_Performance.py provides the formatting and content for the station performance dashboard.

2_Subscribe_For_Alerts.py provides a mockup of what the subscription page could look like.

Functions for creating charts are located within charts.py.

Functions for querying the database are located within data_access.py.

This PR closes #55 and #98.

### CODE REVIEW

File: dashboard/Dockerfile
Comment (aribasyeda) [inline dashboard/Dockerfile:1]: You may need to add the line `COPY .streamlit/config.toml .streamlit/config.toml` so our config settings are kept

File: dashboard/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+FROM python:latest
+
+WORKDIR /app
+
+COPY Home.py .
+COPY pages ./pages
+COPY data_access.py .
+COPY requirements.txt .
```
Comment (aribasyeda) [inline dashboard/Dockerfile:8]: `COPY requirements.txt .` usually goes before copying other scripts over

File: dashboard/Home.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+
+# Page title
+st.title("RailWatch: The Train Station Tracker")
+
+# Main section
+st.header("Welcome!")
+
+st.markdown(
+    """
```
Comment (aribasyeda) [inline dashboard/Home.py:21]: Are we moving it to another page?

File: dashboard/charts.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import pandas as pd
+import altair as alt
+
+scale = alt.Scale(
+    domain=[0, 0.10, 0.25, 1],
+    range=["green", "yellow", "red", "red"],  # Red for values above 20%
+)
+
+
+def make_donut(input_response: float, input_text: str, input_color: str):
```
Comment (aribasyeda) [inline dashboard/charts.py:12]: Return type hints

File: dashboard/charts.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                legend=None,
+            ),
+        )
+        .properties(width=130, height=130)
+    )
+
+    return plot_bg + plot + text
+
+
+def make_delay_per_hour_chart(df: pd.DataFrame):
```
Comment (aribasyeda) [inline dashboard/charts.py:76]: Missing return type hint

File: dashboard/charts.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                scale=alt.Scale(domain=(0, 0.6)),
+            ),
+            alt.Color("pct_delayed", legend=None, scale=scale),
+        )
+        .properties(width=450, height=320)
+        .configure_axis(grid=False, domain=True)
+    )
+
+
+def make_delay_historical_chart(df: pd.DataFrame):
```
Comment (aribasyeda) [inline dashboard/charts.py:98]: Missing return type hints

File: dashboard/charts.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+                axis=alt.Axis(title="Date", format="%d %b"),
+            ),
+            alt.Y("delays:Q", axis=alt.Axis(title="Volume of delays")),
+        )
+        .properties(width=400, height=300)
+        .configure_axis(grid=False, domain=True)
+    )
+
+
+def make_cancellation_historical_chart(df: pd.DataFrame):
```
Comment (aribasyeda) [inline dashboard/charts.py:116]: Return type hints

File: dashboard/data_access.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    database=ENV["DB_NAME"],
+    user=ENV["DB_USER"],
+    password=ENV["DB_PASS"],
+    host=ENV["DB_HOST"],
+    port=ENV["DB_PORT"],
+)
+
+
+cur = conn.cursor()
+
```
Comment (aribasyeda) [inline dashboard/data_access.py:20]: I think you've forgotten to add this to `if __name__ == "__main__":`

File: dashboard/data_access.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+def get_station_names() -> list[str]:
```
Comment (aribasyeda) [inline dashboard/data_access.py:22]: Missing doctring

File: dashboard/data_access.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+            ON arrivals.station_id = stations.station_id
+        WHERE stations.station_name LIKE %s
+        """,
+        (station_name,),
+    )
+
+    return cur.fetchone()[0]
+
+
+def get_total_cancellations_for_station(station_name: str) -> int:
```
Comment (aribasyeda) [inline dashboard/data_access.py:49]: Missing doctring

File: dashboard/data_access.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+            ON cancellations.station_id = stations.station_id
+        WHERE stations.station_name LIKE %s
+        """,
+        (station_name,),
+    )
+
+    return cur.fetchone()[0]
+
+
+def get_total_delays_for_station(threshold: int, station_name: str) -> int:
```
Comment (aribasyeda) [inline dashboard/data_access.py:64]: Missing doctring

File: dashboard/data_access.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+    cur.execute(
+        """
+        SELECT COUNT(*)
+        FROM arrivals
+        LEFT JOIN stations
+            ON arrivals.station_id = stations.station_id
+        WHERE stations.station_name LIKE %s
+        AND EXTRACT(EPOCH FROM actual_arrival - scheduled_arrival)/60 > %s
```
Comment (aribasyeda) [inline dashboard/data_access.py:72]: If you prefer, you can store all of your SQL queries in separate Python file (such as `sql_queries.py`) and storing them as variables can indeed make your code more organised and easier to read

File: dashboard/data_access.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        """,
+        (station_name, threshold),
+    )
+
+    return cur.fetchone()[0]
+
+
+def get_arrivals_for_station(station_name: str) -> pd.DataFrame:
```
Comment (aribasyeda) [inline dashboard/data_access.py:80]: Missing docstring

File: dashboard/data_access.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+        WHERE station_name = %s""",
+        (station_name,),
+    )
+
+    return pd.DataFrame(cur.fetchall())
+
+
+def get_delay_breakdown_for_station(
+    station_name: str, delay_threshold: int, days_delta: int = 30
+):
```
Comment (aribasyeda) [inline dashboard/data_access.py:100]: Missing docstring

File: dashboard/Home.py (hunk 2 above)
Comment (adamski201) [reply dashboard/Home.py:21]: I'll make another pull request for this down the line.

File: dashboard/data_access.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+    """
+
+    params = (station_name, days_delta, delay_threshold, station_name, days_delta)
+
+    cur.execute(query, params)
+
+    return pd.DataFrame(cur.fetchall(), columns=["interval_start", "pct_delayed"])
+
+
+def get_daily_stats(station_name: str, delay_threshold: int, days_delta: int = 30):
```
Comment (aribasyeda) [inline dashboard/data_access.py:133]: Missing docstring

File: dashboard/charts.py (hunk 3 above)
Comment (adamski201) [reply dashboard/charts.py]: Done.

File: dashboard/charts.py (hunk 4 above)
Comment (adamski201) [reply dashboard/charts.py]: Done.

File: dashboard/data_access.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+    params = (station_name, days_delta, delay_threshold, station_name, days_delta)
+
+    cur.execute(query, params)
+
+    return pd.DataFrame(
+        cur.fetchall(), columns=["date", "arrivals", "delays", "cancellations"]
+    )
+
+
+def get_current_incidents(station_name: str):
```
Comment (aribasyeda) [inline dashboard/data_access.py:170]: Missing docstring

File: dashboard/charts.py (hunk 5 above)
Comment (adamski201) [reply dashboard/charts.py:98]: Done.

File: dashboard/charts.py (hunk 6 above)
Comment (adamski201) [reply dashboard/charts.py]: Done.

File: dashboard/pages/1_Station_Performance.py [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+import data_access as data
+from charts import (
+    make_donut,
+    make_delay_per_hour_chart,
+    make_delay_historical_chart,
+    make_cancellation_historical_chart,
+)
+
+
+def calculate_yesterday_percent_delayed(station: str):
```
Comment (aribasyeda) [inline dashboard/pages/1_Station_Performance.py:16]: Missing return type hint

File: dashboard/pages/1_Station_Performance.py [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+    """Returns the percentage of stations delayed yesterday for a station."""
+    threshold = 5
+    n_delayed = data.get_total_delays_for_station(threshold, station)
+    return (100 * n_delayed) / (n_arrivals + n_cancelled)
+
+
+def get_color_for_value(value):
```
Comment (aribasyeda) [inline dashboard/pages/1_Station_Performance.py:23]: Missing type hint

File: dashboard/pages/2_Subscribe_For_Alerts.py [hunk 18]
```diff
@@ -0,0 +1,44 @@
+import streamlit as st
```
Comment (aribasyeda) [inline dashboard/pages/2_Subscribe_For_Alerts.py:1]: I think you've forgotten to add this to `if __name__ == "__main__":`

File: performance_pipeline/.DS_Store
Comment (aribasyeda) [inline performance_pipeline/.DS_Store:1]: You may need to remove this file

**Review by aribasyeda — CHANGES_REQUESTED** [review]
Amazing work Adam 🌟  - left some comments

File: dashboard/data_access.py (hunk 7 above)
Comment (adamski201) [reply dashboard/data_access.py]: Done.

File: dashboard/data_access.py (hunk 8 above)
Comment (adamski201) [reply dashboard/data_access.py]: Done.

File: dashboard/data_access.py (hunk 12 above)
Comment (adamski201) [reply dashboard/data_access.py]: Done.

File: dashboard/data_access.py (hunk 10 above)
Comment (adamski201) [reply dashboard/data_access.py]: Done.

File: dashboard/data_access.py (hunk 9 above)
Comment (adamski201) [reply dashboard/data_access.py]: Done.

File: dashboard/data_access.py [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+        SELECT station_name
+        FROM stations
+        ORDER BY station_name ASC
+        """
+    )
+
+    return [x[0] for x in cur.fetchall()]
+
+
+def get_total_arrivals_for_station(station_name: str) -> int:
```
[the remark this answers is not in the document]
Comment (adamski201) [reply dashboard/data_access.py]: Done.

File: dashboard/data_access.py (hunk 13 above)
Comment (adamski201) [reply dashboard/data_access.py]: Done.

File: dashboard/data_access.py (hunk 14 above)
Comment (adamski201) [reply dashboard/data_access.py]: Done.

File: dashboard/data_access.py (hunk 15 above)
Comment (adamski201) [reply dashboard/data_access.py]: Done.

File: dashboard/pages/1_Station_Performance.py (hunk 16 above)
Comment (adamski201) [reply dashboard/pages/1_Station_Performance.py:19]: Done.

File: dashboard/pages/1_Station_Performance.py (hunk 17 above)
Comment (adamski201) [reply dashboard/pages/1_Station_Performance.py]: Done

File: dashboard/pages/2_Subscribe_For_Alerts.py (hunk 18 above)
Comment (adamski201) [reply dashboard/pages/2_Subscribe_For_Alerts.py]: Page removed as it was a placeholder

File: performance_pipeline/.DS_Store
Comment (adamski201) [reply performance_pipeline/.DS_Store:1]: I believe this was removed rather than added

**Pull request merged**